### PR TITLE
Add email sending account

### LINKED ISSUE #8434: Cognito user pool  email sending account
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

aws_cognito_user_pool email configuration does not currently appear to support adding an email sending account, we currently have to do with via the aws cli / console.

### New or Affected Resource(s)

* aws_cognito_user_pool

### Potential Terraform Configuration

```hcl
resource "aws_cognito_user_pool" "pool" {
  name = "mypool"

  email_configuration {
    reply_to_email_address = "${var.reply_to}"
    source_arn             = "${var.source_arn}"
    email_sending_account  = "DEVELOPER"
  }
}
```

### References

https://docs.aws.amazon.com/cli/latest/reference/cognito-idp/update-user-pool.html

``` json
{
  "SourceArn": "string",
  "ReplyToEmailAddress": "string",
  "EmailSendingAccount": "COGNITO_DEFAULT"|"DEVELOPER"
}
```

### PR DESCRIPTION
**Pull request by SantoDE**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #8434

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```
Adds an option to set the email_sending_account for the cognito email configuration
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSCognitoUserPool_withEmailConfiguration'
=> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSCognitoUserPool_withEmailConfiguration -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSCognitoUserPool_withEmailConfiguration
=== PAUSE TestAccAWSCognitoUserPool_withEmailConfiguration
=== CONT  TestAccAWSCognitoUserPool_withEmailConfiguration
--- FAIL: TestAccAWSCognitoUserPool_withEmailConfiguration (14.04s)
    testing.go:568: Step 0 error: After applying this step, the plan was not empty:

        DIFF:

        UPDATE: aws_cognito_user_pool.pool
          admin_create_user_config.#:                              "1" => "1"
          admin_create_user_config.0.allow_admin_create_user_only: "false" => "false"
          admin_create_user_config.0.invite_message_template.#:    "0" => "0"
          admin_create_user_config.0.unused_account_validity_days: "7" => "7"
          arn:                                                     "arn:aws:cognito-idp:us-west-2:304080159981:userpool/us-west-2_BVFzGb9lf" => "arn:aws:cognito-idp:us-west-2:304080159981:userpool/us-west-2_BVFzGb9lf"
          creation_date:                                           "2019-05-14T09:47:38Z" => "2019-05-14T09:47:38Z"
          device_configuration.#:                                  "0" => "0"
          email_configuration.#:                                   "1" => "0"
          email_configuration.0.email_sending_account:             "COGNITO_DEFAULT" => ""
          email_configuration.0.reply_to_email_address:            "" => ""
          email_configuration.0.source_arn:                        "" => ""
          endpoint:                                                "cognito-idp.us-west-2.amazonaws.com/us-west-2_BVFzGb9lf" => "cognito-idp.us-west-2.amazonaws.com/us-west-2_BVFzGb9lf"
          id:                                                      "us-west-2_BVFzGb9lf" => "us-west-2_BVFzGb9lf"
          lambda_config.#:                                         "0" => "0"
          last_modified_date:                                      "2019-05-14T09:47:38Z" => "2019-05-14T09:47:38Z"
          mfa_configuration:                                       "OFF" => "OFF"
          name:                                                    "terraform-test-pool-4jh1e" => "terraform-test-pool-4jh1e"
          password_policy.#:                                       "1" => "1"
          password_policy.0.minimum_length:                        "8" => "8"
          password_policy.0.require_lowercase:                     "true" => "true"
          password_policy.0.require_numbers:                       "true" => "true"
          password_policy.0.require_symbols:                       "true" => "true"
          password_policy.0.require_uppercase:                     "true" => "true"
          schema.#:                                                "0" => "0"
          sms_configuration.#:                                     "0" => "0"
          user_pool_add_ons.#:                                     "0" => "0"
          verification_message_template.#:                         "1" => "1"
          verification_message_template.0.default_email_option:    "CONFIRM_WITH_CODE" => "CONFIRM_WITH_CODE"
          verification_message_template.0.email_message:           "" => ""
          verification_message_template.0.email_message_by_link:   "" => ""
          verification_message_template.0.email_subject:           "" => ""
          verification_message_template.0.email_subject_by_link:   "" => ""
          verification_message_template.0.sms_message:             "" => ""



        STATE:

        aws_cognito_user_pool.pool:
          ID = us-west-2_BVFzGb9lf
          provider = provider.aws
          admin_create_user_config.# = 1
          admin_create_user_config.0.allow_admin_create_user_only = false
          admin_create_user_config.0.unused_account_validity_days = 7
          arn = arn:aws:cognito-idp:us-west-2:304080159981:userpool/us-west-2_BVFzGb9lf
          creation_date = 2019-05-14T09:47:38Z
          email_configuration.# = 1
          email_configuration.0.email_sending_account = COGNITO_DEFAULT
          email_configuration.0.reply_to_email_address =
          email_configuration.0.source_arn =
          endpoint = cognito-idp.us-west-2.amazonaws.com/us-west-2_BVFzGb9lf
          last_modified_date = 2019-05-14T09:47:38Z
          mfa_configuration = OFF
          name = terraform-test-pool-4jh1e
          password_policy.# = 1
          password_policy.0.minimum_length = 8
          password_policy.0.require_lowercase = true
          password_policy.0.require_numbers = true
          password_policy.0.require_symbols = true
          password_policy.0.require_uppercase = true
          verification_message_template.# = 1
          verification_message_template.0.default_email_option = CONFIRM_WITH_CODE
          verification_message_template.0.email_message =
          verification_message_template.0.email_message_by_link =
          verification_message_template.0.email_subject =
          verification_message_template.0.email_subject_by_link =
          verification_message_template.0.sms_message =
    testing.go:629: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.

        Error: Argument or block definition required: On /var/folders/q3/x4rq343n57x5v8pccndssw3r0000gn/T/tf-test921011959/main.tf line 6: An argument or block definition is required here. To set an argument, use the equals sign "=" to introduce the argument value.

        State: aws_cognito_user_pool.pool:
          ID = us-west-2_BVFzGb9lf
          provider = provider.aws
          admin_create_user_config.# = 1
          admin_create_user_config.0.allow_admin_create_user_only = false
          admin_create_user_config.0.invite_message_template.# = 0
          admin_create_user_config.0.unused_account_validity_days = 7
          arn = arn:aws:cognito-idp:us-west-2:304080159981:userpool/us-west-2_BVFzGb9lf
          creation_date = 2019-05-14T09:47:38Z
          device_configuration.# = 0
          email_configuration.# = 1
          email_configuration.0.email_sending_account = COGNITO_DEFAULT
          email_configuration.0.reply_to_email_address =
          email_configuration.0.source_arn =
          endpoint = cognito-idp.us-west-2.amazonaws.com/us-west-2_BVFzGb9lf
          lambda_config.# = 0
          last_modified_date = 2019-05-14T09:47:38Z
          mfa_configuration = OFF
          name = terraform-test-pool-4jh1e
          password_policy.# = 1
          password_policy.0.minimum_length = 8
          password_policy.0.require_lowercase = true
          password_policy.0.require_numbers = true
          password_policy.0.require_symbols = true
          password_policy.0.require_uppercase = true
          schema.# = 0
          sms_configuration.# = 0
          user_pool_add_ons.# = 0
          verification_message_template.# = 1
          verification_message_template.0.default_email_option = CONFIRM_WITH_CODE
          verification_message_template.0.email_message =
          verification_message_template.0.email_message_by_link =
          verification_message_template.0.email_subject =
          verification_message_template.0.email_subject_by_link =
          verification_message_template.0.sms_message =
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	14.088s
```

Looks like the tests can not clean up. Probably an issue with my account? I also  don't know how to add that case to the AcceptanceTests (as sourceArn is also not included yet), but I did a manual test. See screenshot attached.

image

If someone has a guiding to get that into the Acceptance Tests, I'm happy to help :)

### CODE REVIEW

File: website/docs/r/cognito_user_pool.markdown [hunk 1]
```diff
@@ -65,6 +65,7 @@ The following arguments are supported:
 
   * `reply_to_email_address` (Optional) - The REPLY-TO email address.
   * `source_arn` (Optional) - The ARN of the email source.
+  * `email_sending_account` (Optional) - Instruct Cognito to either use Cognito or SES to send out emails
```
Comment (nywilken) [inline website/docs/r/cognito_user_pool.markdown]:
```suggestion
  * `email_sending_account` (Optional) - Instruct Cognito to either use its built-in functional or Amazon SES to send out emails.
```

**Comment by SantoDE** [comment]
Hey @nywilken ,

thanks for your feedback! the changes are applied. However, the acceptance test is not yet passings. Sadly, it's neither without my modifications so I'm a little lost on how to move forward. Will of course try to figure things out, but a helping hand would be awesome! :)

**Comment by nywilken** [comment]
@SantoDE thanks for the update. I haven't had a chance to revisit since you made updates, but I will take a look in tomorrow my am to see how I can best help with the tests. I have at least confirmed that the acceptance tests are failing on master.

**Comment by SantoDE** [comment]
Hey @nywilken ,

thanks for the heads up! I was out for a couple days, but will have a look at this again tonight :)

**Comment by SantoDE** [comment]
hey @nywilken,

I did some more fiddling around (setting a default for example) but I cant get it to work. Got to admin, I'm a little lost with the whole all of that works together so I guess Im not of a real help anymore :(

**Comment by SantoDE** [comment]
hey again @nywilken,

I tried again to fix this issue but I cant reproduce it outside of the tests. I tried outside with my local test pool and here are the results: 
```
resource "aws_cognito_user_pool" "pool" {
  name = "terraform-example"

  email_configuration {
    email_sending_account = "COGNITO_DEFAULT"
  }
}

```

terraform state:
```
 "resources": {
                "aws_cognito_user_pool.pool": {
                    "type": "aws_cognito_user_pool",
                    "depends_on": [],
                    "primary": {
                        "id": "eu-central-1_grA8qauKu",
                        "attributes": {
                            "admin_create_user_config.#": "1",
                            "admin_create_user_config.0.allow_admin_create_user_only": "false",
                            "admin_create_user_config.0.invite_message_template.#": "0",
                            "admin_create_user_config.0.unused_account_validity_days": "7",
                            "arn": "arn:aws:cognito-idp:eu-central-1:304080159981:userpool/eu-central-1_grA8qauKu",
                            "auto_verified_attributes.#": "0",
                            "creation_date": "2019-06-05T09:04:06Z",
                            "device_configuration.#": "0",
                            "email_configuration.#": "1",
                            "email_configuration.0.email_sending_account": "COGNITO_DEFAULT",
                            "email_configuration.0.reply_to_email_address": "",
                            "email_configuration.0.source_arn": "",
                            "endpoint": "cognito-idp.eu-central-1.amazonaws.com/eu-central-1_grA8qauKu",

```


so it looks like, it's persisting the state just fine but something else is off with the test?

**Comment by nywilken** [comment]
@SantoDE thank you for sticking in there with me.  I didn’t have a moment to follow up  these past two weeks but I’m glad you were able to get the tests passing. I’ll do another review tomorrow to get this merged and lined up for the next release; I’m excited to see the fix. 

Thanks again for the contribution.

File: aws/resource_aws_cognito_user_pool.go [hunk 2]
```diff
@@ -131,6 +131,7 @@ func resourceAwsCognitoUserPool() *schema.Resource {
 				Type:     schema.TypeList,
 				Optional: true,
 				MaxItems: 1,
+				Computed: true,
```
Comment (nywilken) [inline aws/resource_aws_cognito_user_pool.go]: Setting `Computed` for an attribute means that Terraform can't detect drift for unconfigured attributes. In this case the prefer schema change is to add a `DiffSuppressFunc` to ignore the case where Terraform thinks our schema default was a change when it's not, given the default API response.
```suggestion
				DiffSuppressFunc: suppressMissingOptionalConfigurationBlock,
```

File: aws/resource_aws_cognito_user_pool.go [hunk 3]
```diff
@@ -131,6 +131,7 @@ func resourceAwsCognitoUserPool() *schema.Resource {
...
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"reply_to_email_address": {
```
Comment (nywilken) [inline aws/resource_aws_cognito_user_pool.go]: In order to get our testing working I added `""` as a valid value in the validation function.

**Comment by nywilken** [comment]
This was merged manually. Oddly enough GitHub is not showing the correct status and it somehow has the latest changes I tried pushing to your branch which resulted in in a permission denied error locally :man_shrugging:

Thanks again for your patience on this PR and for your contribution.

**Pull request closed without merging**